File: Lab1/src/main/scala/TextData.scala
class TextData(fileName: String, text: String) {
  private var numberOfVowels = 0
  private var numberOfConsonants = 0
  private var numberOfLetters = 0
  private var numberOfSentences = 0
  private var longestWord = ""

  // Analyze text upon initialization
  analyzeText()

  private def analyzeText(): Unit = {
    val words = text.split("\\s+")

    // Find longest word
    words.foreach { word =>
      if (word.length > longestWord.length) {
        longestWord = word
      }
    }

    text.foreach { ch =>
      // Count letters, vowels, and consonants
      if (ch.isLetter) {
        numberOfLetters += 1
        if ("AEIOUaeiou".contains(ch)) {
          numberOfVowels += 1
        } else {
          numberOfConsonants += 1
        }
      } else if (ch == '.' || ch == '!' || ch == '?') { // Check for end of sentence
        numberOfSentences += 1
      }
    }
  }

  def getFileName: String = fileName
  def getText: String = text
  def getNumberOfVowels: Int = numberOfVowels
  def getNumberOfConsonants: Int = numberOfConsonants
  def getNumberOfLetters: Int = numberOfLetters
  def getNumberOfSentences: Int = numberOfSentences
  def getLongestWord: String = longestWord
}

File: lab1/src/main/kotlin/labo1/Task2/TextData.kt
package labo1.Task2


import java.util.*

// TextData class
class TextData(
    //- String fileName
    private val fileName: String = "",
    //- String text
    private var text: String,
    //- int numberOfVowels
    private var numberOfVowels: Int = -1,
    //- int numberOfConsonants
    private var numberOfConsonants: Int = -1,
    //- int numberOfLetters
    private var numberOfLetters: Int = -1,
    //- int numberOfSentences
    private var numberOfSentences: Int = -1,
    //- String longestWord
    private var longestWord: String = ""
) {
    // list of vowels to use in the methods later
    val vowels = listOf('a', 'e', 'i', 'o', 'u')

    // if it were in java, this would have been +TextData(String text)
    constructor(text: String) : this("", text) {
        if (text.isEmpty()) {
            error("Text cannot be an empty string")
        }
    }

    //+ String getFilename()
    public fun getFilename(): String {
        return fileName
    }

    //+ String getText()
    public fun getText(): String {
        return text
    }

    //+ int getNumberOfVowels()
    public fun getNumberOfVowels(): Int {
        if (numberOfVowels == -1) {
            numberOfVowels = text.lowercase().count { it in vowels }
        }
        return numberOfVowels
    }

    //+ int getNumberOfConsonants()
    public fun getNumberOfConsonants(): Int {
        if (numberOfConsonants == -1) {
            numberOfConsonants = text.lowercase().count { it !in vowels && it.isLetter() }
        }
        return numberOfConsonants
    }

    //+ int getNumberOfLetter()
    public fun getNumberOfLetters(): Int {
        if (numberOfLetters == -1) {
            numberOfLetters = text.count { it.isLetter() }
        }
        return numberOfLetters
    }

    //+ int getNumberOfSentences()
    public fun getNumberOfSentences(): Int {
        if (numberOfSentences == -1) {
            numberOfSentences = text.count { it == '.' || it == '!' || it == '?' }
        }
        return numberOfSentences
    }

    //+ String getLongestWord()
    public fun getLongestWord(): String {
        if (longestWord == "") {
            // Use Regex to select only the words
            val words = text.split(Regex("[^\\w]"))
            for (word in words) {
                // Get the longest word
                if (word.length > longestWord.length) {
                    longestWord = word
                }
            }
        }
        return longestWord
    }
}